File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        
        vector<int>prefix(len, 1), suffix(len+1, 1), ans(len);
        
        
        prefix[0] = nums[0];
        suffix[len-1] = nums[len-1];
        
        for(int i=1; i<len; i++){
            prefix[i] = prefix[i-1]*nums[i];
            suffix[len-i-1] = suffix[len-i] * nums[len-i-1];
        }
        
        for(int i=1; i<len; i++){
            ans[i] = prefix[i-1] * suffix[i+1];
        }

        ans[0] = suffix[1];
        
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] product = new int[nums.length];
        
        int val=1;
        product[0]=1;
        for(int i=1; i<nums.length; i++){
            product[i] = product[i-1]* nums[i-1];
        }
        
        for(int i=nums.length-2; i>=0; i--){
            val = val * nums[i+1];
            product[i] = product[i] * val;
        }
        return product;
    }
}